Compute padding percentage correctly for position-area

fantasai opened the pull request:
#### b59ae7a025c9154ab39f3c745f0c333c761d0b98
<pre>
Compute padding percentage correctly for position-area
<a href="https://bugs.webkit.org/show_bug.cgi?id=294256">https://bugs.webkit.org/show_bug.cgi?id=294256</a>
<a href="https://rdar.apple.com/153636536">rdar://153636536</a>

Reviewed by Alan Baradlay.

Percentage padding is supposed to be resolved against the inline size of the
containing block. The position-area property modifies the containing block,
but we weren&apos;t accounting for this in in our padding calculations.

This patch fixes this by updating containingBlockLogicalWidthForContent
(and containingBlockLogicalHeightForContent) to call PositionedLayoutConstraints,
which handles all of the complications for out-of-flow containing block sizing.

It also splits PositionedLayoutConstraints constructor into two pieces, so that
we aren&apos;t doing extra work when we don&apos;t need to, and does a bunch of additional
clean up in PositionedLayoutConstraints to make all of this clearer and cleaner.

Changes include:
 - reorganizing the API into methods that can be called with/without computeInsets()
 - rename m_marginPercentageBasis to m_containingInlineAxis and give it a getter
 - introduce new computeInsets() method to split out second half of constructor
 - store renderer and selfAxis internally, and delete them from method parameters
 - updating all call sites accordingly

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-percents-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-percents-001-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-percents-001.html: Added.

Test for percentage margin/padding/insets in both axes, with mixed writing modes.

* Source/WebCore/rendering/PositionedLayoutConstraints.cpp:
(WebCore::PositionedLayoutConstraints::PositionedLayoutConstraints):

Split the constructor&apos;s logic, keeping just the CB computation.
Store renderer/selfAxis parameters.
Rename local variables for consistency with new containingInlineSize() method.

(WebCore::PositionedLayoutConstraints::computeInsets):

New method for the second half of the constructor (inset/margin computation).

(WebCore::PositionedLayoutConstraints::captureInsets
(WebCore::PositionedLayoutConstraints::captureGridArea):
(WebCore::PositionedLayoutConstraints::captureAnchorGeometry):
(WebCore::PositionedLayoutConstraints::resolvePosition const):
(WebCore::PositionedLayoutConstraints::resolveAlignmentShift const):
(WebCore::PositionedLayoutConstraints::computeStaticPosition):
(WebCore::PositionedLayoutConstraints::computeInlineStaticDistance):
(WebCore::PositionedLayoutConstraints::computeBlockStaticDistance):
(WebCore::PositionedLayoutConstraints::fixupLogicalTopPosition const):

Delete parameters in favor of new data members.

* Source/WebCore/rendering/PositionedLayoutConstraints.h:
(WebCore::PositionedLayoutConstraints::containingSize const):
(WebCore::PositionedLayoutConstraints::containingInlineSize const):
(WebCore::PositionedLayoutConstraints::containingAxis const):
(WebCore::PositionedLayoutConstraints::container const):
(WebCore::PositionedLayoutConstraints::defaultAnchorBox const):
(WebCore::PositionedLayoutConstraints::alignment const):
(WebCore::PositionedLayoutConstraints::startIsBefore const):
(WebCore::PositionedLayoutConstraints::bordersPlusPadding const):
(WebCore::PositionedLayoutConstraints::marginBefore const):
(WebCore::PositionedLayoutConstraints::marginAfter const):
(WebCore::PositionedLayoutConstraints::insetBefore const):
(WebCore::PositionedLayoutConstraints::insetAfter const):
(WebCore::PositionedLayoutConstraints::marginBeforeValue const):
(WebCore::PositionedLayoutConstraints::marginAfterValue const):

Reorganize the API definition to clarify which methods can be called without
calling computeInsets().
Add new m_selfAxis and m_renderer members to cache these inputs.
Rename m_marginPercentageBasis to m_containingInlineSize.

* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::containingBlockLogicalWidthForContent const):
(WebCore::RenderBox::containingBlockLogicalHeightForContent const):

Update to use PositionedLayoutConstraints for out-of-flow boxes.

(WebCore::RenderBox::computePositionedLogicalWidth const):
(WebCore::RenderBox::computePositionedLogicalHeight const):
(WebCore::RenderBox::computePositionedLogicalWidthReplaced const):
(WebCore::RenderBox::computePositionedLogicalHeightReplaced const):
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::overflowsInsetModifiedContainingBlock):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::sortPositionOptionsIfNeeded):

Update PositionedLayoutConstraints call sites.

Canonical link: <a href="https://commits.webkit.org/297275@main">https://commits.webkit.org/297275@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ceaadcb7c0814b5705beafde332e566ee3d71586

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111142 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30808 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21239 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117173 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61410 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113104 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31489 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39390 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84493 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114089 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25156 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100060 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64939 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24501 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18204 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60993 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94536 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18268 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120143 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38191 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28383 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93429 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38567 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96335 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93260 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23762 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38340 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16083 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34176 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38080 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43557 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37745 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41078 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39447 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay reviewed (approved):
I am curious as to why rename/code-shuffle is being mixed with actual functionality changes here. It's hard to figure out what to focus on and in case of a revert we would lose these nice cleanups.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-percents-001-ref.html:
Needs a -expected.html clone if you want this to run in the WebKit test runner

fantasai replied:
I had one, but apparently failed to commit it. :/

fantasai committed:
Compute padding percentage correctly for position-area
https://bugs.webkit.org/show_bug.cgi?id=294256
rdar://153636536

Reviewed by Alan Baradlay.

Percentage padding is supposed to be resolved against the inline size of the
containing block. The position-area property modifies the containing block,
but we weren't accounting for this in in our padding calculations.

This patch fixes this by updating containingBlockLogicalWidthForContent
(and containingBlockLogicalHeightForContent) to call PositionedLayoutConstraints,
which handles all of the complications for out-of-flow containing block sizing.

It also splits PositionedLayoutConstraints constructor into two pieces, so that
we aren't doing extra work when we don't need to, and does a bunch of additional
clean up in PositionedLayoutConstraints to make all of this clearer and cleaner.

Changes include:
 - reorganizing the API into methods that can be called with/without computeInsets()
 - rename m_marginPercentageBasis to m_containingInlineAxis and give it a getter
 - introduce new computeInsets() method to split out second half of constructor
 - store renderer and selfAxis internally, and delete them from method parameters
 - updating all call sites accordingly

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-percents-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-percents-001-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-percents-001.html: Added.

Test for percentage margin/padding/insets in both axes, with mixed writing modes.

* Source/WebCore/rendering/PositionedLayoutConstraints.cpp:
(WebCore::PositionedLayoutConstraints::PositionedLayoutConstraints):

Split the constructor's logic, keeping just the CB computation.
Store renderer/selfAxis parameters.
Rename local variables for consistency with new containingInlineSize() method.

(WebCore::PositionedLayoutConstraints::computeInsets):

New method for the second half of the constructor (inset/margin computation).

(WebCore::PositionedLayoutConstraints::captureInsets
(WebCore::PositionedLayoutConstraints::captureGridArea):
(WebCore::PositionedLayoutConstraints::captureAnchorGeometry):
(WebCore::PositionedLayoutConstraints::resolvePosition const):
(WebCore::PositionedLayoutConstraints::resolveAlignmentShift const):
(WebCore::PositionedLayoutConstraints::computeStaticPosition):
(WebCore::PositionedLayoutConstraints::computeInlineStaticDistance):
(WebCore::PositionedLayoutConstraints::computeBlockStaticDistance):
(WebCore::PositionedLayoutConstraints::fixupLogicalTopPosition const):

Delete parameters in favor of new data members.

* Source/WebCore/rendering/PositionedLayoutConstraints.h:
(WebCore::PositionedLayoutConstraints::containingSize const):
(WebCore::PositionedLayoutConstraints::containingInlineSize const):
(WebCore::PositionedLayoutConstraints::containingAxis const):
(WebCore::PositionedLayoutConstraints::container const):
(WebCore::PositionedLayoutConstraints::defaultAnchorBox const):
(WebCore::PositionedLayoutConstraints::alignment const):
(WebCore::PositionedLayoutConstraints::startIsBefore const):
(WebCore::PositionedLayoutConstraints::bordersPlusPadding const):
(WebCore::PositionedLayoutConstraints::marginBefore const):
(WebCore::PositionedLayoutConstraints::marginAfter const):
(WebCore::PositionedLayoutConstraints::insetBefore const):
(WebCore::PositionedLayoutConstraints::insetAfter const):
(WebCore::PositionedLayoutConstraints::marginBeforeValue const):
(WebCore::PositionedLayoutConstraints::marginAfterValue const):

Reorganize the API definition to clarify which methods can be called without
calling computeInsets().
Add new m_selfAxis and m_renderer members to cache these inputs.
Rename m_marginPercentageBasis to m_containingInlineSize.

* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::containingBlockLogicalWidthForContent const):
(WebCore::RenderBox::containingBlockLogicalHeightForContent const):

Update to use PositionedLayoutConstraints for out-of-flow boxes.

(WebCore::RenderBox::computePositionedLogicalWidth const):
(WebCore::RenderBox::computePositionedLogicalHeight const):
(WebCore::RenderBox::computePositionedLogicalWidthReplaced const):
(WebCore::RenderBox::computePositionedLogicalHeightReplaced const):
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::overflowsInsetModifiedContainingBlock):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::sortPositionOptionsIfNeeded):

Update PositionedLayoutConstraints call sites.

Canonical link: https://commits.webkit.org/297275@main
